scenes/game/ai_player.gd

```diff
@@ -845,168 +903,145 @@ func determine_choose_opponent_card_to_discard_options(card_ids : Array):
 			possible_actions.append(DiscardContinuousBoostAction.new(card.id, true))
 	if limitation not in ["mine", "in_opponent_space"]:
-		for card in opponent.continuous_boosts:
-			if can_pick_discard_continuous(me, opponent, card, limitation, boost_name_restriction):
+		for card in game_opponent.continuous_boosts:
+			if can_pick_discard_continuous(card, limitation, boost_name_restriction):
 				possible_actions.append(DiscardContinuousBoostAction.new(card.id, false))
-	update_ai_state(game_logic, me, opponent)
 	return ai_policy.pick_discard_continuous(possible_actions, game_state)
 
-func can_pick_discard_continuous(me : LocalGame.Player, opponent : LocalGame.Player, card, limitation, boost_name_restriction):
+func can_pick_discard_continuous(card, limitation, boost_name_restriction):
 	if 'cannot_discard' in card.definition['boost'] and card.definition['boost']['cannot_discard']:
 		return false
 	if limitation == "in_opponent_space":
-		var card_location = me.get_boost_location(card.id)
-		if card_location == -1 or not opponent.is_in_location(card_location):
+		var card_location = game_player.get_boost_location(card.id)
+		if card_location == -1 or not game_opponent.is_in_location(card_location):
 			return false
 		if boost_name_restriction and card.definition['boost']['display_name'] != boost_name_restriction:
 			return false
 	return true
 
-func pick_discard_opponent_gauge(game_logic : LocalGame, my_id : Enums.PlayerId) -> DiscardGaugeAction:
-	var me = game_logic._get_player(my_id)
-	var opponent = game_logic._get_player(game_logic.get_other_player(my_id))
+func pick_discard_opponent_gauge() -> DiscardGaugeAction:
+	game_state.update()
 	var possible_actions = []
-	for card in opponent.gauge:
+	for card in game_opponent.gauge:
 		possible_actions.append(DiscardGaugeAction.new(card.id))
-	update_ai_state(game_logic, me, opponent)
 	return ai_policy.pick_discard_opponent_gauge(possible_actions, game_state)
 
-func pick_name_opponent_card(game_logic : LocalGame, my_id : Enums.PlayerId, normal_only : bool, can_use_own_reference : bool = false) -> NameCardAction:
-	var me = game_logic._get_player(my_id)
-	var opponent = game_logic._get_player(game_logic.get_other_player(my_id))
+func pick_name_opponent_card(normal_only : bool, can_use_own_reference : bool = false) -> NameCardAction:
+	game_state.update()
 	var possible_actions = []
-	for i in range(0, opponent.deck_list.size(), 2):
+	for i in range(0, game_opponent.deck_list.size(), 2):
 		# Skip every other card to avoid dupes.
-		var card = opponent.deck_list[i]
+		var card = game_opponent.deck_list[i]
 		if normal_only and card.definition['type'] != "normal":
 			continue
 		possible_actions.append(NameCardAction.new(card.id))
 	if can_use_own_reference:
 		# TODO: cull this down so normals aren't included twice.
-		for i in range(0, me.deck_list.size(), 2):
+		for i in range(0, game_player.deck_list.size(), 2):
 			# Skip every other card to avoid dupes.
```

Comment:
This is wrong for some characters isn't it?
Who has odd numbered amounts of cards? At least Gordeau has a singleton.

Reply:
Happy Chaos has a bunch of 3-sets. I'll slap a TODO on this for now since it's not relevant to the stated purpose of this PR.